From a5980fe7374b8efbbe00e0455be853315c0b6128 Mon Sep 17 00:00:00 2001
From: Richard Yu <richard.yu@mitrastar.com.tw>
Date: Sun, 8 Jun 2025 13:51:04 +0800
Subject: [PATCH 1/2] fronthaul: refresh_vaps_info: fix crash due to map
 modification during iteration

A segmentation fault in check_enabled_vap() was caused by handle_vap()
removing entries from m_radio_info.available_vaps while iterating over it.

To prevent undefined behavior, iteration now operates on a copy of the
map, allowing safe modification of the original during the process.

https://prplfoundationcloud.atlassian.net/browse/PPM-3341

Signed-off-by: Richard Yu <richard.yu@mitrastar.com.tw>
---
 common/beerocks/bwl/whm/base_wlan_hal_whm.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)
 mode change 100644 => 100755 common/beerocks/bwl/whm/base_wlan_hal_whm.cpp

diff --git a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
old mode 100644
new mode 100755
index 243c0f8046..5bd59448e6
--- a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
@@ -911,7 +911,8 @@ bool base_wlan_hal_whm::refresh_vaps_info(int id)
             LOG(ERROR) << "can't find vap_id " << id;
         }
     } else {
-        for (const auto &vap : saved_vaps) {
+        auto saved_vaps_copy = saved_vaps;
+        for (const auto &vap : saved_vaps_copy) {
             if (!vap.second.bss.empty()) {
                 handle_vap(vap.first, vap.second);
             }
-- 
GitLab


From 8e7265e5d60baa5cd094bcbc5a92c045b08072bf Mon Sep 17 00:00:00 2001
From: Richard Yu <richard.yu@mitrastar.com.tw>
Date: Tue, 10 Jun 2025 07:17:49 +0000
Subject: [PATCH 2/2] fronthaul: refresh_vaps_info: clarify reason for copying
 map

Adds a comment above saved_vaps_copy to explain why a snapshot of
m_radio_info.available_vaps is needed during iteration.

The clarification helps avoid confusion by explicitly stating that
modifying the original map during iteration would otherwise lead to
undefined behavior.

Suggested-by: Volodymyr Pavlenko
Signed-off-by: Richard Yu <richard.yu@mitrastar.com.tw>
---
 common/beerocks/bwl/whm/base_wlan_hal_whm.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
index 5bd59448e6..553304103c 100755
--- a/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
+++ b/common/beerocks/bwl/whm/base_wlan_hal_whm.cpp
@@ -911,6 +911,7 @@ bool base_wlan_hal_whm::refresh_vaps_info(int id)
             LOG(ERROR) << "can't find vap_id " << id;
         }
     } else {
+        // Copy to safely iterate while allowing modifications in the original map
         auto saved_vaps_copy = saved_vaps;
         for (const auto &vap : saved_vaps_copy) {
             if (!vap.second.bss.empty()) {
-- 
GitLab

